### PR TITLE
Add podName and svcName length validate

### LINKED ISSUE #1074: Is vcjob name's length limited?
> /kind bug
> /kind feature

When I create a name i found it same no rules for name length.

And the code same only judge the 0.

Guys, do you have the same issue?

**Environment**:
- Volcano Version:v0.4.0
- Kubernetes version (use `kubectl version`):1.17.3
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):unbuntu 18.04
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by mikechengwei**
Fix #1074 
# What is changed and how does it work?
Add podName and svcName length validate

Signed-off-by: mikechengwei 842725815@qq.com

### CODE REVIEW

File: pkg/webhooks/admission/jobs/validate/admit_job.go [hunk 1]
```diff
@@ -274,3 +277,18 @@ func validateTaskTemplate(task v1alpha1.TaskSpec, job *v1alpha1.Job, index int)
 
 	return ""
 }
+
+func validatePodNameLength(template v1.PodTemplateSpec, job *v1alpha1.Job, index int) string {
```
Comment (Thor-wl) [inline pkg/webhooks/admission/jobs/validate/admit_job.go:281]: I'm ok about the validation. Perhaps strict validation just as k8s resource rule is better.

**Comment by Thor-wl** [comment]
Please make CI happy.

File: pkg/webhooks/admission/jobs/validate/admit_job.go (hunk 1 above)
Comment (mikechengwei) [reply pkg/webhooks/admission/jobs/validate/admit_job.go:281]: I have rename method to `validateK8sPodNameLength` and `validateK8sSvcNameLength`.What do you think?

Comment (Thor-wl) [reply pkg/webhooks/admission/jobs/validate/admit_job.go]: > I have rename method to `validateK8sPodNameLength` and `validateK8sSvcNameLength`.What do you think?

OK. Can you give a stricter rule validation? For example, the least length/start character of name/character category of name.

Comment (mikechengwei) [reply pkg/webhooks/admission/jobs/validate/admit_job.go]: I have optimize code.

File: pkg/webhooks/admission/jobs/validate/admit_job.go [hunk 2]
```diff
@@ -32,6 +30,8 @@ import (
 	k8score "k8s.io/kubernetes/pkg/apis/core"
 	k8scorev1 "k8s.io/kubernetes/pkg/apis/core/v1"
 	k8scorevalid "k8s.io/kubernetes/pkg/apis/core/validation"
+	"strings"
```
Comment (shinytang6) [inline pkg/webhooks/admission/jobs/validate/admit_job.go:33]: The correct order of imported packages should be system package => external package => internal package

File: pkg/webhooks/admission/jobs/validate/admit_job.go [hunk 3]
```diff
@@ -274,3 +277,18 @@ func validateTaskTemplate(task v1alpha1.TaskSpec, job *v1alpha1.Job, index int)
 
 	return ""
 }
+
+func validateK8sPodNameLength(template v1.PodTemplateSpec, job *v1alpha1.Job, index int) string {
```
Comment (shinytang6) [inline pkg/webhooks/admission/jobs/validate/admit_job.go:281]: why not use podName as param directly

File: pkg/webhooks/admission/jobs/validate/admit_job.go (hunk 2 above)
Comment (mikechengwei) [reply pkg/webhooks/admission/jobs/validate/admit_job.go:33]: ok

File: pkg/webhooks/admission/jobs/validate/admit_job.go (hunk 3 above)
Comment (mikechengwei) [reply pkg/webhooks/admission/jobs/validate/admit_job.go]: done .

File: pkg/webhooks/admission/jobs/validate/admit_job.go [hunk 4]
```diff
@@ -32,6 +31,7 @@ import (
 	k8score "k8s.io/kubernetes/pkg/apis/core"
 	k8scorev1 "k8s.io/kubernetes/pkg/apis/core/v1"
 	k8scorevalid "k8s.io/kubernetes/pkg/apis/core/validation"
+	jobhelpers "volcano.sh/volcano/pkg/controllers/job/helpers"
```
Comment (shinytang6) [inline pkg/webhooks/admission/jobs/validate/admit_job.go:34]: this should be placed in inner package

Comment (shinytang6) [reply pkg/webhooks/admission/jobs/validate/admit_job.go:34]: And Ci is failed, l think you should run `make unit-test` or `make e2e` locally to check it

File: pkg/webhooks/admission/jobs/validate/admit_job.go [hunk 5]
```diff
@@ -274,3 +278,17 @@ func validateTaskTemplate(task v1alpha1.TaskSpec, job *v1alpha1.Job, index int)
...
 }
+
+func validateK8sPodNameLength(podName string) string {
+	if errMsgs := validation.IsDNS1123Label(podName); len(errMsgs) > 0 {
+		return fmt.Sprintf("create pod with name %s validate failed %v;", podName, errMsgs)
+	}
+	return ""
+}
+
+func validateK8sSvcNameLength(job *v1alpha1.Job) string {
```
Comment (Thor-wl) [inline pkg/webhooks/admission/jobs/validate/admit_job.go:289]: It suddenly occurred to me that these two validation may be found in k8s.io

Comment (mikechengwei) [reply pkg/webhooks/admission/jobs/validate/admit_job.go:289]: validation.IsDNS1123Label  method ？

Comment (mikechengwei) [reply pkg/webhooks/admission/jobs/validate/admit_job.go:289]: I think we just need to  validate podname and svcname by validation.IsDNS1123Label method

File: pkg/webhooks/admission/jobs/validate/admit_job.go (hunk 2 above)
Comment (mikechengwei) [reply pkg/webhooks/admission/jobs/validate/admit_job.go:33]: I follow this order, verify-fmt.sh will report an error.

File: pkg/webhooks/admission/jobs/validate/admit_job.go (hunk 5 above)
Comment (Thor-wl) [reply pkg/webhooks/admission/jobs/validate/admit_job.go:290]: Please go fmt midified files and make CI happy

Comment (Thor-wl) [reply pkg/webhooks/admission/jobs/validate/admit_job.go:290]: > I think we just need to validate podname and svcname by validation.IsDNS1123Label method

`IsQualifiedName` method may be a good choice ?

Comment (mikechengwei) [reply pkg/webhooks/admission/jobs/validate/admit_job.go:290]: Ok, i have updated

Comment (mikechengwei) [reply pkg/webhooks/admission/jobs/validate/admit_job.go:290]: Now, do you have any other questions? @Thor-wl

File: pkg/webhooks/admission/jobs/validate/admit_job.go [hunk 6]
```diff
@@ -152,13 +153,16 @@ func validateJobCreate(job *v1alpha1.Job, reviewResponse *v1beta1.AdmissionRespo
...
+			msg += err.Error() + fmt.Sprintf(" valid events are %v, valid actions are %v",
 				getValidEvents(), getValidActions())
 		}
-
+		podName := jobhelpers.MakePodName(job.Name, task.Template.Name, index)
+		msg += validateK8sPodNameLength(podName)
 		msg += validateTaskTemplate(task, job, index)
 	}
 
+	msg += validateK8sSvcNameLength(job)
```
Comment (william-wang) [inline pkg/webhooks/admission/jobs/validate/admit_job.go:164]: I am confused about the func name "validateK8sSvcNameLength".  No services here treatment in func, "validateJobName" seems better.

**Review by william-wang — CHANGES_REQUESTED** [review]

File: pkg/webhooks/admission/jobs/validate/admit_job.go (hunk 6 above)
Comment (mikechengwei) [reply pkg/webhooks/admission/jobs/validate/admit_job.go:164]: Because we use job name as pod svc name .

Comment (mikechengwei) [reply pkg/webhooks/admission/jobs/validate/admit_job.go]: Ok,i have change the method name.

**Comment by mikechengwei** [comment]
@william-wang Do you have time lgtm?

**Comment by Thor-wl** [comment]
> @william-wang Do you have time lgtm?

Don't worry, william is often busy with work at daytime. He will review community at night.

**Comment by mikechengwei** [comment]
> > @william-wang Do you have time lgtm?
> 
> Don't worry, william is often busy with work at daytime. He will review community at night.
OK , thanks for reminding

**Review by william-wang — APPROVED** [review]
/lgtm
/approve

**Pull request merged**